gtsfm/densify/patchmatchnet_data.py

```diff
@@ -0,0 +1,259 @@
+                        depths[key_b].append(b_z)
+
+                    # if both cameras are valid cameras
+                    if key_a >= 0 and key_b >= 0:
+                        # calculate score for track_i in the pair views (cam_a, cam_b)
+                        score_a_b = piecewise_gaussian(
+                            xPa=self._camera_centers[i_a] - w_x, xPb=self._camera_centers[i_b] - w_x
+                        )
+                        # sum up pair scores for each track_i
+                        pair_scores[key_a, key_b] += score_a_b
+                        pair_scores[key_b, key_a] += score_a_b
+
+        # sort pair scores, for i-th row, choose the largest (num_views-1) scores, the corresponding views are selected
+        #   as (num_views-1) source views for i-th reference view.
+        pairs = np.argsort(pair_scores, axis=1)[:, -self._num_views + 1 :][:, ::-1]
+
+        # filter out depth outliers and calculate depth ranges
+        depth_ranges = np.zeros((self._num_images, 2))
+        for i in range(self._num_images):
+            depths_sorted = sorted(depths[i])
+            depth_ranges[i, 0] = depths_sorted[int(len(depths[i]) * 0.01)]
+            depth_ranges[i, 1] = depths_sorted[int(len(depths[i]) * 0.99)]
+
+        return pairs, depth_ranges
+
+    def __len__(self) -> int:
+        """Get the number of images
+
+        Returns:
+            length of image dictionary's keys
+        """
+        return self._num_images
+
+    def __getitem__(self, index: int) -> Dict[str, Any]:
+        """Get inference data to PatchmatchNet
+        produce data containing _num_views images, the first images is the reference image
+
+        Args:
+            index: index of yield test data, the reference image ID of the test data is _keys[index]
+
+        Returns:
+            Dictionary containing:
+                "idx" test image index: int
+                "imgs" source and reference images: (num_views, image_channel, image_h, image_w)
+                "proj_matrices" projection matrices: (num_views, 4, 4)
+                "depth_min" minimum depth: int
+                "depth_max" maximum depth: int
+                "filename" output filename pattern: string
+        """
+        ref_key = self._patchmatchnet_idx_to_camera_idx[index]
+        src_keys = [self._patchmatchnet_idx_to_camera_idx[src_index] for src_index in self._pairs[index]]
+
+        cam_keys = [ref_key] + src_keys
+
+        imgs: List[np.ndarray] = [[] for _ in range(self._num_stages)]
+        proj_mats: List[np.ndarray] = [[] for _ in range(self._num_stages)]
+
+        for cam_key in cam_keys:
+            img = self._images[cam_key].value_array
+            np_img = np.array(img, dtype=np.float32) / 255.0
+            np_img = cv2.resize(np_img, (self._w, self._h), interpolation=cv2.INTER_LINEAR)
+            np_img = np_img[: self._cropped_h, : self._cropped_w, :]
```

Comment:
Would be nice to leave an inline comment, like ‘crop from the upper left corner, instead of from the center’

Reply:
Thanks for adding this : - )